website/upgrade-guides/0-13.html.markdown

```diff
@@ -137,6 +150,48 @@ run `terraform init` again to re-run the provider installer.
 
 -> **Action:** Either run [`terraform 0.13upgrade`](/docs/commands/0.13upgrade.html) for each of your modules, or manually update the provider declarations to use explicit source addresses.
 
+The upgrade tool described above only updates references in your configuration.
+The Terraform state also includes references to provider configurations which
+need to be updated to refer to the correct providers.
+
+Terraform will automatically update provider configuration references in the
+state the first time you run `terraform apply` after upgrading, but it relies
+on information in the configuration to understand which provider any
+existing resource belongs to, and so you must run `terraform apply` at least
+once (and accept any changes it proposes) before removing any `resource` blocks
+from your configuration after upgrading.
+
+If you are using Terraform Cloud or Terraform Enterprise with the VCS-driven
+workflow (as oposed to CLI-driven runs), refer to
+[The UI- and VCS-driven Run Workflow](/docs/cloud/run/ui.html) to learn how
+to manually start a run after you select a Terraform v0.13 release for your
+workspace.
+
+If you remove a `resource` block (or a `module` block for a module that
+contains `resource` blocks) before the first `terraform apply`, you may see
+a message like this reflecting that Terraform cannot determine which provider
+configuration the existing object ought to be managed by:
+
+```
+Error: Provider configuration not present
+
+To work with null_resource.foo its original provider configuration at
+provider["registry.terraform.io/-/aws"] is required, but it has been removed.
+This occurs when a provider configuration is removed while objects created by
+that provider still exist in the state. Re-add the provider configuration to
+destroy aws_instance.example, after which you can remove the provider
+configuration again.
+```
+
+In this specific upgrade situation the problem is actually the missing
+`resource` block rather than the missing `provider` block: Terraform would
+normally refer to the configuration to see if this resource has an explicit
+`provider` argument that would override the default strategy for selecting
+a provider. If you see the above after upgrading, re-add the resource mentioned
```

Comment:
What do you think about including instructions or a link to the docs for `terraform state replace-provider` as an alternative? (I'm fine with not doing this to avoid muddying the waters).

Reply:
I'm hesitant to introduce `terraform state replace-provider` into the "main" upgrade path (as opposed to the less-common path for folks with in-house providers) because it seems likely to present more opportunities for mistakes and confusion due to it requiring the provider addresses to be spelled out explicitly, and because `terraform state replace-provider` is a "bulk" operation (changing lots of resources all at once) so the impact of a mistake could be quite high.

The `terraform apply` flow here has the advantage of picking up the provider addresses automatically from the configuration, so as long as the user's written their configuration correctly (or, more likely, had it generated by `terraform 0.13upgrade`) the state upgrade will follow suit and produce the correct result.